Introduced Basic256Sha256 security policy based on Basic128Rsa15

bobbthebuilder opened the pull request:
I hope this pull request meets @Infinity95 expectations on a rebase on master to get rid of the redundant commits and being compliant with the new security policy interface that was merged into master just shortly before. This pull request summarizes the effort we made to make Basic256Sha256 encryption usable and working in open62541. The decisive changes we made by @Infinity95. 

@Infinity95 @jpfr please make a quick check that you are able to connect your client with a server that has Basic256Sha256 enabled as well before doing the merge. I'm using the plugin for several days and faced no problems with anonymous logins whereas username/password logins do not work as i reported at the very end of #1587.

jpfr commented:
> whereas username/password logins do not work as i reported at the very end of #1587.

This was fixed in a recent commit.
Can you rebase and make continuous integration pass?

We can then merge.

ChHaa commented:
Tested succesfully with UAExpert, had to tweak CMakeLists a bit to make it compile though

bobbthebuilder commented:
@jpfr In the recent days i was also working on the username/authentification issue when encryption is enabled. Do i understand you right that #1636 "Refactor/Authentification" is already fixing that?
I tested username/password authentification in combination with Basic128Rsa15 from the latest master that already has #1636 merged and got BadIdentityTokenInvalid (I still have the feeling i misunderstood something here).

To get it working i followed the description from the standard, Part 4, Section 7.36 et seq. It states that the password token is encrypted with the AsymmetricEncryptionAlgorithm of the corresponding Security Policy by the client and padded according to 7.362.2 Table 181 that defines the format of the password token. The private key is known to the server and when the activateSession function can access it, the password token can be decrypted, depadded and compared. The username is transmitted in clear text.
#1636 does some nice refactoring but i miss any form of decryption. What do i oversee?

@ChHaa glad to hear that you successfully tested the feature. Can you tell what you had to change in the CMakeLists to get it working?

jpfr commented:
#1636 fixed some obvious warts in the existing access control plugin.
Now the plugin exposes the list of UserTokenPolicies it supports.

The default config adds **all** UserTokenPolicy (anonymous, password, ...) from the access control plugin to all endpoints. An end-user might want to change this. For example by not allowing certain combinations of SecurityPolicy for the channel and UserTokenPolicy for the session.

In order to prevent that passwords are transmitted over an unencrypted SecureChannel, the UserTokenPolicy can again refer to an encryption method. So the password is encrypted even if the SecureChannel is not. It has not been implemented so far.

Cool to see you are interested! The access control plugin is the right place for it. If you need additional API, for example to get a certificate currently not exposed, you are invited to make a proposal for the changes.

I'm going to take a break from this specific topic. Lack of time...
Please refer to @ChHaa and @Infinity95 for details.

ChHaa commented:
Change corresponding lines to e.g.

if(UA_ENABLE_ENCRYPTION)
    set(default_plugin_headers ${default_plugin_headers}
            ${PROJECT_SOURCE_DIR}/plugins/ua_securitypolicy_basic128rsa15.h
            ${PROJECT_SOURCE_DIR}/plugins/ua_securitypolicy_basic256sha256.h)
    set(default_plugin_sources ${default_plugin_sources}
            ${PROJECT_SOURCE_DIR}/plugins/ua_securitypolicy_basic128rsa15.c
            ${PROJECT_SOURCE_DIR}/plugins/ua_securitypolicy_basic256sha256.c)
endif()

bobbthebuilder commented:
@jpfr ahh, now things are clear to me, thank you!
@ChHaa you are right, my mistake, i fixed the CMakeLists.txt accordingly.

I rebased on master again. 
The continuous-intergration errors are weird, i was not able to reproduce them on my box yet.

Infinity95 commented:
You need to add the plugin sources to the testing and fuzzing plugin sources as well.
You probably don't experience these errors because you don't have the unit tests enabled on your machine.

bobbthebuilder commented:
@Infinity95 that's it, thank you!

stuff is ready to be merged i think.

jpfr commented:
Cool! Thanks for the contribution @bobbthebuilder !
Keep up the good work. 😄
